presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBuffer.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.buffer;
+
+import com.facebook.presto.OutputBuffers;
+import com.facebook.presto.execution.StateMachine.StateChangeListener;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface OutputBuffer
+{
+    /**
+     * Gets the current state of this buffer.  This method is guaranteed to not block or acquire
+     * contended locks, but the stats in the info object may be internally inconsistent.
+     */
+    SharedBufferInfo getInfo();
+
+    /**
+     * A buffer is finished once no-more-pages has been set and all buffers have been closed
+     * with an abort call.
+     */
+    boolean isFinished();
+
+    /**
+     * Get the memory utilization percentage.
+     */
+    double getUtilization();
+
+    /**
+     * Add a listener which fires anytime the buffer state changes.
+     */
+    void addStateChangeListener(StateChangeListener<BufferState> stateChangeListener);
+
+    /**
+     * Updates the buffer configuration.
+     */
+    void setOutputBuffers(OutputBuffers newOutputBuffers);
+
+    /**
+     * Gets pages from the output buffer, and acknowledges all pages received from the request
+     * that supplied the specified token.
+     */
+    CompletableFuture<BufferResult> get(TaskId outputId, long token, DataSize maxSize);
+
+    /**
+     * Closes the specified output buffer.
+     */
+    void abort(TaskId outputId);
+
+    /**
+     * Adds a page to an unpartitioned buffer. If no-more-pages has been set, the enqueue
+     * page call is ignored.  This can happen with limit queries.
+     */
+    ListenableFuture<?> enqueue(Page page);
+
+    /**
+     * Adds a page so a specific partition.  If no-more-pages has been set, the enqueue
+     * page call is ignored.  This can happen with limit queries.
+     */
+    ListenableFuture<?> enqueue(int partition, Page page);
+
+    /**
+     * Disallows adding more pages to this output.
+     */
+    void setNoMorePages();
+
+    /**
+     * Destroys the buffer, discarding all pages.
+     */
+    void destroy();
+
+    /**
+     * Fail the buffer, discarding all pages, but blocking readers.
```

Comment:
It would be helpful to add additional comments to explain what happens after readers are blocked.